Provide API for observing the active text editor

jasonrudolph opened the pull request:
### Description of the Change

As the first step toward resolving #14648, this pull request implements the proposal outlined in https://github.com/atom/atom/issues/14648#issuecomment-305017095:

> - Disallow editors in docks. Only allow editors in the workspace center.
> - Change `atom.workspace.getActiveTextEditor()` to return the active text editor (if any) in the workspace center. (This is more consistent with the pre 1.17.0 behavior described in https://github.com/atom/atom/issues/14648#issuecomment-304381899.)
> - Introduce `atom.workspace.observeActiveTextEditor()` as a means of observing the active text editor. (Packages like encoding-selector and grammar-selector _want_ to observe the active text editor, but currently have to rely on observing the active pane item, which is problematic for the reasons described in https://github.com/atom/atom/issues/14648#issuecomment-304391943.)

Once this pull request is merged, I'll then be able to update each of the core packages that currently use `atom.workspace.observeActivePaneItem()` to instead use `atom.workspace.observeActiveTextEditor()` to receive notifications each time a different text editor becomes active. Once that's done, we'll be able to close #14648.

### Alternate Designs

- For packages that need to observe the active text editor (e.g., https://github.com/atom/status-bar/issues/194), we could require those packages to observe all pane containers and watch for a text editor to become active in that pane container. Because observing the active text editor is a common need, I'd rather provide a core API for this functionality and not require each package to implement and maintain this logic.
- Instead of disallowing editors in docks, we could teach `atom.workspace.getActiveTextEditor()` to first check for an active text editor in the active pane container and return it if it exists, and to otherwise return the active text editor (if any) in the workspace center. `atom.workspace.observeActiveTextEditor()` would need to use similar logic to notify observers each time the value of `atom.workspace.getActiveTextEditor()` changes. This approach would likely allow us to resolve https://github.com/atom/status-bar/issues/194, but it feels overly complicated, and it seems like something that is more likely to cause future confusion, surprise, and additional maintenance overhead.

### Why Should This Be In Core?

It's the first step toward resolving #14648.

### Benefits

- Developers will have a core API for getting notified when:
  - A text editor becomes the active text editor
  - There is no longer an active text editor
- The behavior of `atom.workspace.getActiveTextEditor()` will [more closely match its pre-1.17.0 behavior](https://github.com/atom/atom/issues/14648#issuecomment-304381899), which various [packages relied on](https://github.com/atom/find-and-replace/pull/904).
- We answer the question, "Does it make sense to have an editor in a dock?" Answer: No. Atom expects to find editors in the workspace center. Docks are intended to provide supporting tools like a tree view, a debugger, a console. xref: http://blog.atom.io/2017/05/23/docks-deep-dive.html
- It reduces (hopefully 🤞) the overall number of scenarios that need to be considered in Atom core and in packages that interact with editors (e.g., https://github.com/atom/find-and-replace/issues/901#issuecomment-304078622).

### Possible Drawbacks

- By disallowing editors in docks, we reduce Atom's overall flexibility. If someone comes up with a compelling reason for using an editor in a dock, they're out of luck.
- Implementing `getAllowedLocations()` on an item affects where the item opens when using `atom.workspace.open()` and when [dragging an item to a dock in the UI](https://github.com/atom/atom/blob/16e1ef917be66637a58bad5e09c7d546782ee6e8/src/dock.js#L228-L229). However, functions like [`Pane::addItem()`](https://atom.io/docs/api/v1.17.2/Pane#instance-addItem) currently allow you to add the item to any location, regardless of the locations returned by the item's `getAllowedLocations()` function. In other words, there will continue to be ways to programmatically add a text editor to a dock, even though we're discouraging it and considering it to be unsupported behavior.

### Applicable Issues

- #14648
- https://github.com/atom/status-bar/issues/194

### TODO

- [x] Add `TextEditor::getAllowedLocations()` to indicate that editors belong in the workspace center
- [x] Add `Workspace::observeActiveTextEditor()`
- [x] Add API docs for new functions in `Workspace`
- [x] Determine and implement reasonable behavior for existing text editor functions in `Dock`:
  - [x] [`observeTextEditors`](https://github.com/atom/atom/blob/16e1ef917be66637a58bad5e09c7d546782ee6e8/src/dock.js#L387-L400) -- 0b314ac1a98ca06c82fc369fffe4f13fc772abe5
  - [x] [`getTextEditors`](https://github.com/atom/atom/blob/16e1ef917be66637a58bad5e09c7d546782ee6e8/src/dock.js#L586-L591) -- 09495dfc35b1294df0f4785e00d4eb1f8c7e27fc
  - [x] [`getActiveTextEditor`](https://github.com/atom/atom/blob/16e1ef917be66637a58bad5e09c7d546782ee6e8/src/dock.js#L593-L600) -- f8ebd71200af4dcc5cf6fc2d85e68247833d9acc

jasonrudolph committed:
Allow text editors in the workspace center, but not in docks

jasonrudolph committed:
Teach Workspace::getActiveTextEditor() to get item from center

jasonrudolph committed:
Add Workspace::observeActiveTextEditor()

jasonrudolph committed:
Update Workspace specs regarding editors in docks

jasonrudolph committed:
:memo: Fix odd grammar in API docs

jasonrudolph committed:
Deprecate Dock::getActiveTextEditor()

jasonrudolph committed:
:fire: Remove broken Dock::observeTextEditors(callback) fn

We're updating Atom to disallow editors in docks. As a result, we need
to remove Dock::observeTextEditors(callback). Normally, we would
deprecate a method before removing it, but this method is broken and has
never worked:

  atom.workspace.getLeftDock().observeTextEditors(console.log)
  (unknown) Uncaught TypeError: this.paneContainer.getTextEditors is not a function
      at Dock.getTextEditors (/Applications/Atom.app/Contents/Resources/app/src/dock.js:590:37)
      at Dock.observeTextEditors (/Applications/Atom.app/Contents/Resources/app/src/dock.js:396:41)
      at <anonymous>:1:30
  getTextEditors @ <embedded>:29933
  observeTextEditors @ <embedded>:29739
  (anonymous) @ VM1941:1

Since the method is broken, we know that nobody is relying on it.
Instead of deprecating the method, we can just remove it.

jasonrudolph committed:
:fire: Remove broken Dock::getTextEditors() function

We're updating Atom to disallow editors in docks. As a result, we need
to remove Dock::getTextEditors(). Normally, we would deprecate a method
before removing it, but this method is broken and has never worked:

  atom.workspace.getLeftDock().getTextEditors()
  (unknown) Uncaught TypeError: this.paneContainer.getTextEditors is not a function
      at Dock.getTextEditors (/Applications/Atom.app/Contents/Resources/app/src/dock.js:590:37)
      at <anonymous>:1:30

Since the method is broken, we know that nobody is relying on it.
Instead of deprecating the method, we can just remove it.

jasonrudolph committed:
:memo: Add API docs for onDidChangeActiveTextEditor(callback)

jasonrudolph committed:
:memo: Add API docs for observeActiveTextEditor(callback)

maxbrunsfeld commented on src/workspace.js:
```diff
@@ -422,6 +423,17 @@ module.exports = class Workspace extends Model {
+
+    if (paneContainer === this.getCenter()) {
+      const activeTextEditorChanged =
+        (item instanceof TextEditor) || (this.activeTextEditor instanceof TextEditor)
```
I think it'd be nice to consolidate these `item instanceof TextEditor` checks into one. Also, I would think we would not need to check `this.activeTextEditor instanceof TextEditor`, because `this.activeTextEditor` is always either a text editor or null.

jasonrudolph replied:
Thanks, @maxbrunsfeld. :bow:

I pushed up a few updates that attempt to incorporate these suggestions. If these updates are different from what you had in mind, just let me know.

marlazinger200 replied:
1

maxbrunsfeld reviewed (commented):
This looks great! I just suggested one change.

jasonrudolph committed:
:art: Refactor: Introduce explaining variable

xref: https://github.com/atom/atom/pull/14695#discussion_r119721719

jasonrudolph committed:
:art: Replace instanceof check with nullness check

xref: https://github.com/atom/atom/pull/14695#discussion_r119721719

jasonrudolph committed:
:art: Reduce ternaries

jasonrudolph committed:
:bug: Initialize active editor state correctly after reload

Fixes the following bug:

1. Open Atom
2. Open a file
3. Observe the file's encoding in the status bar
4. Reload Atom
5. Close the file
6. Observe that the closed file's encoding is still present in the
status bar

This bug occured because the reload did not deserialize/serialze the
workspace's active text editor state. As a result, when closing the
text editor in step 5, we failed to notify observers that there is no
longer an active text editor.

jasonrudolph commented:
5b61c0a949e70d258b5ce2657dc9a1f5507d1878 fixes a bug that I uncovered related to serialization/deserialization.

@maxbrunsfeld: This is my first time touching any serialization/deserialization code. Would you mind taking a look at 5b61c0a949e70d258b5ce2657dc9a1f5507d1878 and letting me know if there's anything I should change?

jasonrudolph committed:
Remove unnecessary serialization

jasonrudolph committed:
Test deserialization in terms of user-observable functionality

- Rework serialization/deserialization test
- Move simulateReload function so that it can be used in multiple
describe blocks

jasonrudolph committed:
Always invoke callback with current active text editor

This makes observeActiveTextEditor consistent with observers like
observeActivePaneItem, which always invoke the callback with the current
value, regardless of whether that value is undefined or not.

jasonrudolph committed:
Teach Workspace::reset() to properly reset hasActiveTextEditor

ungb commented:
@jasonrudolph I found a few small bugs.  

1.  If you select tree view you can't change the line endings

![focustree](https://cloud.githubusercontent.com/assets/5624870/26817477/9bbd3fa2-4a4b-11e7-99fa-499e73009819.gif)

2. says file uses undefined line ending (could be unrelated to your pr)
![image](https://cloud.githubusercontent.com/assets/5624870/26817497/b9e56a2c-4a4b-11e7-9548-73179a01c56e.png)

3.  If you select the tree view last, close and reopen Atom the file ending is missing on start up, hard to repro.  Also doesn't repro if you completely exit out of Atom.  If you have the file selected last, it seems to work correctly.
![missingle](https://cloud.githubusercontent.com/assets/5624870/26817664/48e8deb6-4a4c-11e7-8838-cd421aff3254.gif)

ungb commented:
issue number 2 I listed, looks like it was a recent pr that was merged. see https://github.com/atom/line-ending-selector/pull/42

jasonrudolph commented:
> NOTE: 1,2 are issues that already exist on master so they aren't caused by your pr. The only one I think is no.3, but currently you can't see any of the items in the status bar. PR is a HUGE improvement over what's currently out.

@ungb: Thanks for finding these issues. :zap: https://github.com/atom/line-ending-selector/commit/2ff4ee3f833c6dc8489518aa4815e20238011f4d should resolve # 1 and # 3.

lgeiger commented:
Thanks for adding this API! This is a lot more convenient than doing:
```js
atom.workspace.getCenter().observeActivePaneItem(item => {
  const editor = item instanceof TextEditor ? item : null;
  ...
})
```
May I suggest to backport this API to Atom 1.18 beta if it can be cherry-picked cleanly?

jasonrudolph commented:
> May I suggest to backport this API to Atom 1.18 beta if it can be cherry-picked cleanly?

@lgeiger: Yes. 😸  Before the 1.18 stable release, we're hoping to publish a new 1.18-beta release including this fix.

marlazinger200 commented:
1
